### PR TITLE
Implement PairSemigroup

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Signed-off-by: Mihai Budiu <mbudiu@vmware.com>
This is needed for average, for example

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
How can coverage decrease when I implement one function and add tests for it *and* 1 more?

File: src/algebra/mod.rs
Comment (ryzhyk) [inline src/algebra/mod.rs:569]: These two bounds are redundant.

**Review by ryzhyk — CHANGES_REQUESTED** [review]

**Review by ryzhyk — APPROVED** [review]

**Comment by mbudiu-vmw** [comment]
Do we run in CI randon python scripts we pull from the internet? This sounds dangerous to me.
The bench-skylake-2x failed with this error:
```
 python3 gh-pages/_scripts/compare_nexmark.py
  File "/home/github-runner/actions-runner/_work/database-stream-processor/database-stream-processor/gh-pages/_scripts/compare_nexmark.py", line 158
    df_compare['Peak RSS diff'] = current_results['allocstats_after_peak_rss'] -
                                                                                ^
SyntaxError: invalid syntax
```
Because of this failure I cannot merge this PR.

**Comment by ryzhyk** [comment]
This is @gz's script.  I think he made a typo while fixing something else earlier today.  This PR can be merged, the bug is unrelated.

**Comment by mbudiu-vmw** [comment]
I could not find the script in the repository.
My merge button is disabled.

**Comment by mbudiu-vmw** [comment]
Turns out I also need to derive Clone for it...

**Comment by mbudiu-vmw** [comment]
It is too complicated to fight with the CI of this repo, I will add this to the SQL runtime library

**Pull request closed without merging**
